src/thunderbots/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,272 @@
+    // not going to converge to this new passer point if they've already been converging
+    // to another passer point for a while
+    passes_to_optimize = generatePasses(num_passes_to_optimize.value());
+}
+
+std::optional<Pass> PassGenerator::getBestPassSoFar()
+{
+    // Take ownership of the best_known_pass for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
+
+    return best_known_pass;
+}
+
+void PassGenerator::setTargetRegion(Rectangle area)
+{
+    // Take ownership of the target_region for the duration of this function
+    std::lock_guard<std::mutex> target_region_lock(target_region_mutex);
+
+    this->target_region = std::move(area);
+}
+
+
+PassGenerator::~PassGenerator()
+{
+    // Set this flag so pass_generation_thread knows to end (also making sure to
+    // properly take and give ownership of the flag)
+    in_destructor_mutex.lock();
+    in_destructor = true;
+    in_destructor_mutex.unlock();
+
+    // Join to pass_generation_thread so that we wait for it to exit before destructing
+    // the thread object. If we do not wait for thread to finish executing, it will
+    // call `std::terminate` when we deallocate the thread object and kill our whole
+    // program
+    pass_generation_thread.join();
+}
+
+void PassGenerator::continuouslyGeneratePasses()
+{
+    // Take ownership of the in_destructor flag so we can use it for the conditional
+    // check
+    in_destructor_mutex.lock();
+    while (!in_destructor)
+    {
+        // Give up ownership of the in_destructor flag now that we're done the
+        // conditional check
+        in_destructor_mutex.unlock();
+
+        optimizePasses();
+        pruneAndReplacePasses();
+        saveBestPath();
+
+        // Yield to allow other threads to run. This is particularly important if we
+        // have this thread and another running on one core
+        std::this_thread::yield();
+
+        // Take ownership of the `in_destructor` flag so we can use it for the conditional
+        // check
+        in_destructor_mutex.lock();
+    }
+}
+
+void PassGenerator::optimizePasses()
+{
+    // The objective function we minimize in gradient descent to improve each pass
+    // that we're optimizing
+    const auto objective_function =
+        [this](std::array<double, NUM_PARAMS_TO_OPTIMIZE> pass_array) {
+            Pass pass = convertArrayToPass(pass_array);
+            return ratePass(pass);
+        };
+
+    // Run gradient descent to optimize the passes to for the requested number
+    // of iterations
+    // NOTE: Parallelizing this `for` loop would probably be safe and potentially more
+    //       performant
+    for (Pass& pass : passes_to_optimize)
+    {
+        auto pass_array =
+            optimizer.maximize(objective_function, convertPassToArray(pass),
+                               number_of_gradient_descent_steps_per_iter.value());
```

Comment:
Maybe add a TODO to test the performance of some different sets of hyperparameters for the optimization being used

Reply:
Great point, we should definitely invest considerable effort in tuning, could significantly improve results. Opened https://github.com/UBC-Thunderbots/Software/issues/388 to tune all the various hyperparameters after we've finished implementation.